### PR TITLE
implemented reflection at pi/2

### PR DESCRIPTION
**Pull request by aaditya-thakkar**
@rwst, please have a look at this patch. This tries to solve http://trac.sagemath.org/ticket/20139 and #149 . If anything needs to be added, please suggest me. :-)

### CODE REVIEW

**Comment by aaditya-thakkar** [comment]
A small bug is there in the previous commit. It only checks for angles greater than `pi/2` in the condition `if (not is_ex_the_function(res, cos))`, which should be logically `if (not is_ex_the_function(res, cos) && not is_ex_the_function(_ex_1*res, cos))` on line 982, otherwise `sec(104*pi/105)` would give `-1/cos(1/105*pi)` instead of giving `-sec(1/105*pi)`. Same is the case on line 831 too.

**Comment by aaditya-thakkar** [comment]
Here is the fix for this. :)

File: ginac/inifcns_trig.cpp [hunk 1]
```diff
@@ -176,6 +176,24 @@ static ex sin_eval(const ex & x)
...
+                // Reflection at Pi/2
+                else {
+                        if (is_exactly_a<numeric>(coef_pi)) {
+                                const numeric c = ex_to<numeric>(coef_pi);
+                                if (c.is_rational()) {
+                                        const numeric num = c.numer();
+                                        const numeric den = c.denom();
+                                        const numeric rm = num.mod(den);
+                                        
+                                        if (2*rm.to_int() > den.to_int()) {
```
Comment (rwst) [inline ginac/inifcns_trig.cpp]: You will get overflow if you reduce big integers to int, and indeed:
```
sage: sin(pi*(2^31-1)/2^31)
sin(1/2147483648*pi)
sage: sin(pi*(2^33-1)/2^33)
sin(8589934591/8589934592*pi)
```

File: ginac/inifcns_trig.cpp [hunk 2]
```diff
@@ -722,6 +776,18 @@ static ex cot_evalf(const ex & x, PyObject* parent)
 
 static ex cot_eval(const ex & x)
 {
+        std::function<bool (const ex&)> has_pi =
```
Comment (rwst) [inline ginac/inifcns_trig.cpp]: I think we should no longer use several identical lambdas but one static function. Lambdas are also slower I heard. Could you please do a separate PR with thzis change?

Comment (rwst) [inline ginac/inifcns_trig.cpp]: Or rather, a separate commit in this PR.

File: ginac/inifcns_trig.cpp (hunk 1 above)
Comment (aaditya-thakkar) [inline ginac/inifcns_trig.cpp]: may be use `to_long()` instead of `to_int()`?

Comment (rwst) [inline ginac/inifcns_trig.cpp]: No. Numerics can be of type MPZ, MPQ, and PYOBJECT. All arithmetic operations on them are available, and because numerics are polymorphic you don't need to know which types they are most of the time.

**Comment by aaditya-thakkar** [comment]
I have opened a new PR as you said. Closing this one. :)
